src/storage/src/table/cell_based_table.rs

```diff
@@ -185,22 +187,103 @@ impl<S: StateStore> CellBasedTable<S> {
 
     pub async fn batch_write_rows(
         &mut self,
-        rows: Vec<(Row, Option<Row>)>,
+        buffer: BTreeMap<Row, RowOp>,
         epoch: u64,
     ) -> StorageResult<()> {
         let mut batch = self.keyspace.state_store().start_write_batch();
         let mut local = batch.prefixify(&self.keyspace);
-        for (pk, cell_values) in rows {
-            let arrange_key_buf =
-                serialize_pk(&pk, self.pk_serializer.as_ref().unwrap()).map_err(err)?;
-            let bytes = self
-                .cell_based_row_serializer
-                .serialize(&arrange_key_buf, cell_values, &self.column_ids)
-                .map_err(err)?;
-            for (key, value) in bytes {
-                match value {
-                    Some(val) => local.put(key, StorageValue::new_default_put(val)),
-                    None => local.delete(key),
+        let ordered_row_serializer = self.pk_serializer.as_ref().unwrap();
+        for (pk, row_op) in buffer {
+            let arrange_key_buf = serialize_pk(&pk, ordered_row_serializer).map_err(err)?;
+            match row_op {
+                RowOp::Insert(row) => {
+                    let bytes = self
+                        .cell_based_row_serializer
+                        .serialize(&arrange_key_buf, row, &self.column_ids)
+                        .map_err(err)?;
+                    for (key, value) in bytes {
+                        local.put(key, StorageValue::new_default_put(value))
+                    }
+                }
+                RowOp::Delete(old_value) => {
+                    // TODO(wcy-fdu): only serialize key on deletion
+                    let bytes = self
+                        .cell_based_row_serializer
+                        .serialize(&arrange_key_buf, old_value, &self.column_ids)
+                        .map_err(err)?;
+                    for (key, _) in bytes {
+                        local.delete(key);
+                    }
+                }
+                RowOp::Update((old_row, new_row)) => {
+                    let delete_bytes = self
```

Comment:
Do we need to serialize the full old_row? Looks like we only need pk and cell id for old_row.

Reply:
Yes, that's an optimization for the future.